sqlflow_models/dbscan.py

```diff
@@ -0,0 +1,179 @@
+        # initiate with the first data
+        clusters = [neighbors_list[0] + [0]]
+        for i in range(1, len(neighbors_list)):
+            # for other data in the neighbour list
+            # check if the data has an intersected cluster inside the result list
+            # merge the list and append it to the result
+            list_of_intersected_cluster = []
+            new_cluster = neighbors_list[i] + [i]
+            for cluster_num in range(len(clusters)):
+                if self.intersect(neighbors_list[i],
+                                  clusters[cluster_num]):
+                    list_of_intersected_cluster.append(clusters[cluster_num])
+                    new_cluster = new_cluster + clusters[cluster_num]
+
+            # if the data is a new cluster / no intersected clusters
+            if not list_of_intersected_cluster:
+                clusters.append(neighbors_list[i] + [i])
+            else:
+                clusters.append(list(set(new_cluster)))
+                # delete the merged clusters
+                for old_cluster in list_of_intersected_cluster:
+                    clusters.remove(old_cluster)
+        return clusters
+
+    def labelling(self, data, clusters):
+        cluster_labels = [self.OUTLIER] * len(data)
+        for i in range(len(self.clusters)):
+            for j in range(len(self.clusters[i])):
+                cluster_labels[self.clusters[i][j]] = i
+        return cluster_labels
+
+    def fit(self, X):
+        distance_matrix = squareform(pdist(X))
+        # compute the neighbors
+        neighbors = self.compute_neighbors(distance_matrix)
+        # clustering
+        self.clusters = self.generate_clusters(neighbors)
+        # filter out clusters with neighbors < minpts
+        self.clusters = list(filter(lambda x: len(x) >= self.minpts,
+                                    self.clusters))
+        # labelling
+        self.labels_ = np.array(self.labelling(X, self.clusters))
+
+        return self
+
+    def _split_dataset(self, dataset):
+        pass
+
+    # do custom training here, parameter "dataset" is a tf.dataset type representing the input data.
+    def sqlflow_train_loop(self, dataset, useIrisDemo=True, epochs=1, verbose=0):
+        if useIrisDemo == True:
+            from sklearn import datasets, metrics
+            iris = datasets.load_iris()  # <class 'sklearn.utils.Bunch'>
+            x_df = iris.data  # (150, 4) numpy.ndarray float64
+            y_df = iris.target
+            self.fit_predict(x_df)
+            print("DBSCAN (minpts=10, eps=0.4): %f" %
+                  purity_score(y_df, self.labels_))
+        else:
+            x_df, y_df = self._split_dataset(dataset)
+            self.fit_predict(x_df)
+            print("DBSCAN (minpts=10, eps=0.4): %f" %
+                  purity_score(y_df, self.labels_))
+'''
+if __name__ == '__main__':
+    from sklearn.datasets.samples_generator import make_blobs
```

Comment:
Can we remove the comment code and testing this model in the `test_db_scan.py`? A referenced test script: https://github.com/sql-machine-learning/models/blob/develop/tests/test_arima_with_stl_decomposition.py

Reply:
Comment code removed and rename the test_db_scan.py.